### PR TITLE
Implement Wasm GC spec in IPInt

### PR DESCRIPTION
**Pull request by danlliu**
#### b15346e3a1fed37b9370efc28732fe731f973c08
<pre>
Implement Wasm GC spec in IPInt
<a href="https://bugs.webkit.org/show_bug.cgi?id=283986">https://bugs.webkit.org/show_bug.cgi?id=283986</a>
<a href="https://rdar.apple.com/140866114">rdar://140866114</a>

Reviewed by Yusuke Suzuki.

To catch up to LLInt, we need to support the Wasm GC
and typed function reference proposals in IPInt.

* JSTests/wasm/ipint-tests/ipint-test-array-simpleinit.js: Added.
(import.as.assert.from.string_appeared_here.module):
(test):
* JSTests/wasm/ipint-tests/ipint-test-default-ref-local.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.result.i32.local.funcref.local.0.ref.is_null.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-gc-starter.js: Added.
(import.as.assert.from.string_appeared_here.module):
(test):
* JSTests/wasm/ipint-tests/ipint-test-struct-packed.js: Added.
(import.as.assert.from.string_appeared_here.module):
(test):
* JSTests/wasm/ipint-tests/ipint-test-struct-set.js: Added.
(import.as.assert.from.string_appeared_here.module):
(test):
* JSTests/wasm/ipint-tests/ipint-test-struct-simple.js: Added.
(import.as.assert.from.string_appeared_here.module):
(test):
* JSTests/wasm/spec-harness/async_index.js:
(eq_ref):
(assert_invalid_custom):
(reinitializeRegistry.get promise_test):
(reinitializeRegistry):
* JSTests/wasm/spec-harness/sync_index.js:
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter.cpp:
(JSC::IPInt::initialize):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addLocal):
(JSC::Wasm::IPIntGenerator::addRefI31):
(JSC::Wasm::IPIntGenerator::addI31GetS):
(JSC::Wasm::IPIntGenerator::addI31GetU):
(JSC::Wasm::IPIntGenerator::addArrayNew):
(JSC::Wasm::IPIntGenerator::addArrayNewData):
(JSC::Wasm::IPIntGenerator::addArrayNewElem):
(JSC::Wasm::IPIntGenerator::addArrayNewFixed):
(JSC::Wasm::IPIntGenerator::addArrayNewDefault):
(JSC::Wasm::IPIntGenerator::addArrayGet):
(JSC::Wasm::IPIntGenerator::addArraySet):
(JSC::Wasm::IPIntGenerator::addArrayLen):
(JSC::Wasm::IPIntGenerator::addArrayFill):
(JSC::Wasm::IPIntGenerator::addArrayCopy):
(JSC::Wasm::IPIntGenerator::addArrayInitElem):
(JSC::Wasm::IPIntGenerator::addArrayInitData):
(JSC::Wasm::IPIntGenerator::addStructNew):
(JSC::Wasm::IPIntGenerator::addStructNewDefault):
(JSC::Wasm::IPIntGenerator::addStructGet):
(JSC::Wasm::IPIntGenerator::addStructSet):
(JSC::Wasm::IPIntGenerator::addRefTest):
(JSC::Wasm::IPIntGenerator::addRefCast):
(JSC::Wasm::IPIntGenerator::addAnyConvertExtern):
(JSC::Wasm::IPIntGenerator::addExternConvertAny):
(JSC::Wasm::IPIntGenerator::addBranchNull):
(JSC::Wasm::IPIntGenerator::addBranchCast):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:
* Tools/lldb/debug_ipint.py:

Canonical link: <a href="https://commits.webkit.org/287403@main">https://commits.webkit.org/287403@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e430f6cf1f229a19ae8b445273b45c77c9f0d754

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32737 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83973 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30503 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81496 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6652 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62078 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19961 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82430 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52134 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72309 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42386 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49478 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26494 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28904 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72452 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85377 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78555 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6641 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4625 "Found 1 new test failure: imported/w3c/web-platform-tests/wasm/webapi/esm-integration/script-src-blocks-wasm.tentative.sub.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70329 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6805 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68184 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69576 "Found 2 new API test failures: /TestWebKit:WebKit.EnumerateDevicesCrash, /TestWebKit:WebKit.LoadAlternateHTMLStringWithNonDirectoryURL (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13622 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12487 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100894 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12285 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6593 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24615 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6508 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9983 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8306 "Built successfully") | | |

### CODE REVIEW

File: Source/JavaScriptCore/wasm/WasmIPIntGenerator.h [hunk 1]
```diff
@@ -247,6 +260,77 @@ enum class UIntBytecode: uint8_t {
...
+};
+
+struct ArrayInitElemMetadata {
+    uint32_t elemSegmentIndex;
+    uint8_t length;
+};
+
+struct RefTestCastMetadata {
+    int32_t typeIndex;
+    bool allowNull;
```
Comment (justinmichaud) [inline Source/JavaScriptCore/wasm/WasmIPIntGenerator.h]: This isn't needed by the asm code, right? Can the cpp helper just look this up from the wasm type index? Alternatively, can we arrange these structs to be more compact?

**Review by justinmichaud — COMMENTED** [review]
Overall LGTM! Nice

File: Source/JavaScriptCore/wasm/WasmIPIntGenerator.h (hunk 1 above)
Comment (danlliu) [reply Source/JavaScriptCore/wasm/WasmIPIntGenerator.h]: Yup! Turns out, our `shouldNegate` flag is solely just for LLInt to handle `br_on_cast`/`br_on_cast_fail` without adding a custom instruction for those... I'm working on the patches now but now we should only need a type index and length. Thanks for the catch :)

**Review by Constellation — APPROVED** [review]
r=me

**Pull request merged**